smt/surrogate_models/krg_based.py

```diff
@@ -381,6 +396,250 @@ def _predict_variances(self, x):
+                        k.append(x[j][i])
+                        i = i + 1
+                    y = np.zeros(np.shape(k))
+                    y[np.argmax(k)] = 1
+                    x[j][i0:i] = y
+                    ##Categorial : The biggest level is selected.
+        return x
+
+    def _transform_vartype(self, dim):
+        """
+        This function unfold vartype list to a coded array with
+        0 for continuous variables, 1 for integers and n>1 for each 
+        level of the n-th categorical variable.
+        Each level correspond to a new continuous dimension.
+        
+        --------
+        Arguments
+        ---------
+        vartype : list
+            The type of the each dimension  (cont, int or cate).
+        dim : int
+            The number of dimension
+        Returns
+        -------
+        vartype : np.ndarray
+            The type of the each dimension. 
+        """
+        vartype = self.options["vartype"]
+        if vartype is None:
+            vartype = np.zeros(dim)
+        if isinstance(vartype, list):
+            temp = []
+            ind_cate = 2
+            for i in vartype:
+                if i == "cont":
+                    temp.append(0)
+                    ##new continuous dimension : append 0
+                elif i == "int":
+                    temp.append(1)
+                    ##new integer dimension : append 1
+                elif i[0] == "cate":
+                    for j in range(i[1]):
+                        temp.append(ind_cate)
+                    ##For each level
+                    ##new categorical dimension : append n
+                    ind_cate = ind_cate + 1
+                else:
+                    raise Exception ("type_error")
+            temp = np.array(temp)
+            vartype = temp
+        ## Assign 0 to continuous variables, 1 for int and n>1 for each
+        # categorical variable.
+
+        return vartype
+
+    def _relax_limits(self, xlimits):
+        """
+        This function unfold xlimits to add contiuous dimensions
+        Each level correspond to a new continuous dimension in [0,1].
+        Integer dimensions are relaxed continuously.
+        
+        --------
+        Arguments
+        ---------
+        vartype : list
+        The type of the each original dimension  (cont, int or cate).
+        xlimits : np.ndarray
+        The bounds of the each original dimension and their labels .
+        Returns
+        -------
+        xlimits : np.ndarray
+        The bounds of the each original dimension  (cont, int or cate).
+        """
+        
+        #Continuous optimization : do nothing
+        if (self.options["vartype"] is None ):
+            return xlimits
+        
+        self.options["vartype"]= self._transform_vartype(self.options["vartype"])
+        xlim=xlimits
+        vt=self.options["vartype"]
+        
+        #continuous or integer => no cate
+        if (isinstance(xlim[0][0], np.float64)) :
```

Comment:
I am not sure about this test on xlimits... What about xlim[1][0] when there are several dimensions?